LTSupportAutomotive/LTOBD2Command.m

```diff
@@ -117,35 +123,115 @@ -(BOOL)isCAN
     {
         NSMutableString* ms = [NSMutableString string];
-        NSArray<NSNumber*>* bytes = [self.cookedResponse objectForKey:ecu];
+        NSArray<NSNumber*>* bytes = [self.responsePayload objectForKey:ecu];
         [bytes enumerateObjectsUsingBlock:^(NSNumber * _Nonnull byte, NSUInteger idx, BOOL * _Nonnull stop) {
             [ms appendFormat:@"%02X", byte.unsignedIntValue];
         }];
         [ma addObject:ms];
     }
     
+    // default implementation returns a single NSString
     return [ma componentsJoinedByString:@", "];
 }
 
+
+/**
+ Units string used when building the formattedResponse
+ Default units is empty string.
+ Subclasses can override to define their own units.
+*/
+-(NSString*)units
+{
+    return @"";
+}
+
+/**
+ Format string used when building the formattedResponse
+ Default format is simple string concatentation
+ Subclasses can override to define their own formatting for integer and double value types.
+*/
+-(NSString*)format
+{
+    return @"%@";
+}
+
+/**
+ Declares the type of the decodedResponse object.
+   - String
+   - Integer
+   - Double
+ Default responseDataType is a String.
+ Subclasses can override to configure their own responseDataType.
+*/
+-(LTResponseDataType)responseDataType
+{
+    return LTString;
+}
+
+/**
+ Default formattedResponse implementation
+ Uses the format string to format the decodeResponse object and then appends the units string.
+ Subclasses can override to customise their own formattedResponse string.
+ However, most commands do not need to customise this function, and can just configure format and units strings and provide an impl of decodeResponse.
+*/
+-(NSString*)formattedResponse
+{
+    if ( !self.responsePayload )
+    {
+        return LTStringLookupWithPlaceholder(@"OBD2_NO_DATA", @"N/A");
+    }
+    
+    NSObject* decodedResponse = self.decodeResponse;
+    
+    LOG( @"decodedResponse: '%@'", decodedResponse );
+    
+    NSString* responseString;
+    
+    if (self.responseDataType == LTString) {
+        responseString = [NSString stringWithFormat:self.format, decodedResponse];
+    } else if (self.responseDataType == LTInteger) {
+        int decodedIntValue = ((NSNumber*) decodedResponse).intValue;
+        responseString = [NSString stringWithFormat:self.format, decodedIntValue ];
+    } else if (self.responseDataType == LTDouble) {
+        double decodedDoubleValue = ((NSNumber*) decodedResponse).doubleValue;
+        responseString = [NSString stringWithFormat:self.format, decodedDoubleValue ];
+    } else {
+        LOG( @"ERROR - UNHANDLED ResponseDataType: '%@'", self.responseDataType );
+    }
+    
+    LOG( @"responseString after format: '%@': '%@'", self.format, responseString );
+
+    if (self.units) {
+        responseString = [NSString stringWithFormat: @"%@" UTF8_NARROW_NOBREAK_SPACE @"%@", responseString, self.units];
+        LOG( @"responseString with units: '%@'", responseString );
+    }
+    
+    return responseString;
+}
+
```

Comment:
The introduction of a `unit` concept is a great change in principle, something I would have wanted to implement right from the start, but never got around to actually doing it – instead we have to live with that inflexible `formattedResponse` that conveys little semantics.

That said though, if you are to introduce a `unit` concept, I'd rather see it utilizing the concepts from `Foundation`, which already contains classes such as `NSUnit`, `NSMeasurement`, and `NSUnitFormatter`.

Reply:
Sounds like a good idea. I will look at your suggestion.

Reply:
Mickey, I have updated my changes to use NSUnit, NSMeasurement, and NSUnitFormatter. Will send a PR soon.